Blog: keycloak (#122)
JankariTech/blog: src/assets/keycloak/keycloak.md

---

> title: keycloak

Reviewer: try to find a title that is more descriptive and attractive for the reader :)

authorName: Amrita Shrestha
authorAvatar: https://avatars.githubusercontent.com/u/54478846?s=100&v=4
authorLink: https://github.com/amrita-shrestha
createdAt: March 26, 2024
tags: Keycloak
banner: https://blog.jankaritech.com/src/assets/keycloak/images/keycloak.png
---

## Keycloak

> Keycloak is an [open-source](https://github.com/keycloak/keycloak) project for Red Hat SSO. It provides `Identity and Access Management` solution designed to secure application services.

Reviewer: I think the link should go with the name, not with the term "open-source"

> Keycloak is an [open-source](https://github.com/keycloak/keycloak) project for Red Hat SSO. It provides `Identity and Access Management` solution designed to secure application services.

Reviewer: ```suggestion
Keycloak is an [open-source](https://github.com/keycloak/keycloak) project for Red Hat SSO. It provides an `Identity and Access Management` solution designed to secure application services.
```

or 
```suggestion
Keycloak is an [open-source](https://github.com/keycloak/keycloak) project for Red Hat SSO. It provides `Identity and Access Management` solutions designed to secure application services.
```
if this is singular, I think this needs an article. Check with the English teacher or with @phil-davis :)

Additionally, it enables users to authenticate through various identity providers and use fine-grained permissions for regulating access to Software as a Service (SaaS) applications.

> Keycloak alternatives are Auth0, Zluri, Microsoft Azure Active Directory, Okta, WSO2 Identity Server.

Reviewer: ```suggestion
Keycloak alternatives are Auth0, Zluri, Microsoft Azure Active Directory, Okta or WSO2 Identity Server.
```

Keycloak facilitates the creation of a user database with customizable roles and groups, offering functionalities such as user management, registration, and password policy enforcement.

> This makes it a comprehensive Identity and Access Management (IAM) solution for contemporary applications. Keycloak's API enables the integration of all these features into your application

Reviewer: usually the abbreviation is introduced the first time a term is mentioned in a text

without the need for additional coding.

> Despite its extensive features, Keycloak does have certain limitations. One of these is the need for a more varied implementation approach.

Reviewer: Usually the drawbacks/limitations are mentioned after pointing out the benefits of using this tool

Additionally, the following are some of the drawbacks associated with Keycloak:
1. `Complex Server Deployment:` The manual process involved in deploying Keycloak on a server can be complicated, potentially affecting overall productivity.
2. `Lack of Comprehensive Support Documentation:` Despite its robust functionalities, Keycloak could benefit from more efficient and extensive support documentation. This would make it easier for users to find suitable solutions to their problems.

## Keycloak Features
![Keycloak features](/src/assets/keycloak/images/keycloak-features.png)

1. `Identity Broker:`
   Identity Broker acts as a middleman, linking various service providers with different identity providers. It establishes trust with external identity providers to utilize their identities for accessing internal services offered by service providers.

>    Imagine textile company,which has two keycloak setups: one for customer and another for internal employees. If you want your employees to access external services without creating new accounts on customer keycloak, you can use the customer Keycloak

Reviewer: ```suggestion
   Imagine a textile company, which has two keycloak setups: one for customer and another for internal employees. If you want your employees to access external services without creating new accounts on customer keycloak, you can use the customer Keycloak
```

Reviewer: sometimes you write Keycloak with capital `K` and somtimes with small `k` . Make sure to be consistent through your text. Names are in general written with capital letter unless it's part of the Corporate Design to write it with small letters

   as the mediator and the customer Keycloak as the source in this scenario.

> 2. `User federation:`

Reviewer: ```suggestion
2. `User Federation:`
```
There are two ways for spelling titles in English: either all words are capitalized or only the first word (unless there are names). Try to be consistent in which style you apply throughout your blog. You can also ask the English teacher for recommendations on this.

>    Keycloak provides the ability to integrate with existing LDAP or Active Directory severs. It also support implementation of your own provider such as a relational database.

Reviewer: ```suggestion
   Keycloak provides the ability to integrate with existing LDAP or Active Directory servers. It also supports implementation of your own provider such as a relational database.
```

3. `Standard protocols:`

>    Keycloak base on three standard protocols i.e OpenID Connect, OAuth2.0, and SAML.

Reviewer: ```suggestion
   Keycloak is based on three standard protocols i.e., OpenID Connect, OAuth2.0, and SAML.
```

4. `Password Policies:`

>    Keycloak contain different password policies i.e HashAlgorithm, Hashing Iterations, Digits, Lowercase Characters, Regular Expression and so on.

Reviewer: ```suggestion
   Keycloak contain different password policies i.e., HashAlgorithm, Hashing Iterations, Digits, Lowercase Characters, Regular Expression and so on.
```

Reviewer: comma before and in lists is optional, however try to be consistent throughout your text if you use a comma there or not, see also https://www.grammarly.com/blog/comma-before-and and/or chat with your English teacher about this :)

5. `Single-Sign On:`
   Keycloak facilitates a seamless login experience where users only need to sign in once to access multiple applications, eliminating the need for repeated logins.
   This not only streamlines user experience but also simplifies identity management for administrators.

6. `Social Login:`

>    Keycloak has built-in support for Google, Github, Facebook which help us to use social identity providers.

Reviewer: ```suggestion
   Keycloak has built-in support for Google, Github, and Facebook which help us to use social identity providers.
```
